### PR TITLE
RTS2DScene and RTS3DInterfaceScene

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+    m_needUpdate = false;
+    VertexBufferClass::WriteLockClass lock(m_vertexBufferCircle, 0);
+    VertexFormatXYZDUV1 *vertices = (VertexFormatXYZDUV1 *)lock.Get_Vertex_Array();
+    float z = 0.0f;
+    float x = 0.02f;
+    int diffuse = m_diffuse + 0x7F000000;
+    int count = m_numTriangles;
+    float angle = 0.0f;
+    float deltaAngle = 6.2831855f / (float)count;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp]: pi * 2

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        indices[1] = i + 1;
+        indices[2] = i + 2;
+        indices += 3;
+    }
+
+    m_vertexBufferCircle =
+        new DX8VertexBufferClass(DX8_FVF_XYZDUV1, 3 * m_numTriangles, DX8VertexBufferClass::USAGE_DEFAULT, 0);
+    m_vertexBufferScreen = new DX8VertexBufferClass(DX8_FVF_XYZDUV1, 6, DX8VertexBufferClass::USAGE_DEFAULT, 0);
+    m_vertexMaterialClass = VertexMaterialClass::Get_Preset(VertexMaterialClass::PRELIT_DIFFUSE);
+    m_shaderClass = ShaderClass(0x984B7);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp:109]: What is 0x984B7 ?

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+int W3DStatusCircle::Update_Circle_VB()
+{
+    if (!m_vertexBufferCircle) {
+        return -1;
+    }
+
+    m_needUpdate = false;
+    VertexBufferClass::WriteLockClass lock(m_vertexBufferCircle, 0);
+    VertexFormatXYZDUV1 *vertices = (VertexFormatXYZDUV1 *)lock.Get_Vertex_Array();
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp:121]: static cast?

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+int W3DStatusCircle::Update_Screen_VB(int diffuse)
+{
+    if (!m_vertexBufferCircle) {
+        return -1;
+    }
+
+    m_needUpdate = false;
+    VertexBufferClass::WriteLockClass lock(m_vertexBufferCircle, 0);
+    VertexFormatXYZDUV1 *vertices = (VertexFormatXYZDUV1 *)lock.Get_Vertex_Array();
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp:179]: static cast?

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+    vertices->x = -1.0;
+    vertices->y = -1.0;
+    vertices->z = 0.0;
+    vertices->diffuse = diffuse;
+    vertices->u1 = 0.0;
+    vertices->v1 = 0.0;
+    vertices++;
+
+    vertices->x = 1.0;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp:175]: -1

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 2 above)
Comment (tomsons26) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp:109]: shader bits, needs a todo to be resolved

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 1 above)
Comment (tomsons26) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp]: should be using the define
and delta_angle

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+#include "dx8indexbuffer.h"
+#include "dx8vertexbuffer.h"
+#include "dx8wrapper.h"
+#include "gamelogic.h"
+#include "globaldata.h"
+#include "scriptengine.h"
+#include "vertmaterial.h"
+
+int W3DStatusCircle::m_diffuse = 0xFF;
+bool W3DStatusCircle::m_needUpdate;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp]: g_

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    int diffuse = m_diffuse + 0x7F000000;
+    int count = m_numTriangles;
+    float angle = 0.0f;
+    float deltaAngle = 6.2831855f / (float)count;
+
+    for (int i = 0; i < count; i++) {
+        for (int j = 0; j < 3; j++) {
+            vertices->z = z;
+
+            if (!j) {
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp]: switch then?

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    if (g_theGameLogic->Is_In_Game() && g_theGameLogic->Get_Game_Mode() != GAME_SHELL) {
+        if (!m_indexBuffer) {
+            Init_Data();
+        }
+
+        if (m_indexBuffer) {
+            bool b = false;
+            Matrix3D m(true);
+
+            if (g_theWriteableGlobalData->m_benchMark) {
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp]: not benchmark, m_showTeamDot, its set in wb by parseShowTeamDot
check if this is in game too as its debug code

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+                DX8Wrapper::Set_Index_Buffer(m_indexBuffer, 0);
+                DX8Wrapper::Set_Vertex_Buffer(m_vertexBufferCircle, 0);
+                b = true;
+                m.Set_Translation(Vector3(0.95f, 0.67f, 0.0f));
+                Matrix4 m2(m);
+                DX8Wrapper::Set_DX8_Transform(D3DTS_WORLD, m2);
+                DX8Wrapper::Draw_Triangles(0, 20, 0, 3 * m_numTriangles);
+            }
+
+            int fade = g_theScriptEngine->Get_Fade();
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp]: is fade the enum?

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+                }
+
+                m.Make_Identity();
+                int fadevalue = (int)(g_theScriptEngine->Get_Fade_Value() * 255.0f);
+                int diffuse = fadevalue | (fadevalue << 8) | (fadevalue << 0x10) | 0xFF000000;
+                Update_Screen_VB(diffuse);
+
+                Matrix4 m2(m);
+                DX8Wrapper::Set_DX8_Transform(D3DTS_WORLD, m2);
+                DX8Wrapper::Set_Shader(ShaderClass(0x94037));
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp:279]: todo on the bits

File: src/platform/w3dengine/client/w3dstatuscircle.cpp [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+#include "dx8wrapper.h"
+#include "gamelogic.h"
+#include "globaldata.h"
+#include "scriptengine.h"
+#include "vertmaterial.h"
+
+int W3DStatusCircle::m_diffuse = 0xFF;
+bool W3DStatusCircle::m_needUpdate;
+
+W3DStatusCircle::W3DStatusCircle() :
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3dstatuscircle.cpp:27]: add m_numTriangles init to 0, comment as bugfix

File: src/game/common/globaldata.h [hunk 12]
```diff
@@ -226,13 +226,12 @@ class GlobalData : public SubsystemInterface
     bool m_displayDebug; // not 100% sure and needs confirming
     bool m_winCursors;
     bool m_constantDebugUpdate;
+    bool m_benchMark;
```
Comment (tomsons26) [inline src/game/common/globaldata.h]: m_showTeamDot and the stats ones need checking, there are in fact two bools, see wb 00687FB2

**Review by tomsons26 — CHANGES_REQUESTED** [review]
just team dot issue remains

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp:109]: fixed.

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 6 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp]: fixed.

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 10 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp:279]: fixed.

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 8 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp]: renamed,.

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 7 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp]: fixed.

File: src/game/common/globaldata.h (hunk 12 above)
Comment (jonwil) [reply src/game/common/globaldata.h]: fixed.

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 11 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp:27]: fixed.

File: src/platform/w3dengine/client/w3dstatuscircle.cpp (hunk 9 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dstatuscircle.cpp]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**